addons/pos_loyalty/static/src/app/models/pos_order.js

```diff
@@ -908,11 +908,17 @@ patch(PosOrder.prototype, {
         return { discountable, discountablePerTax };
     },
     /**
-     * @returns the order's cheapest line
+     * @param {loyalty.reward} reward
+     * @returns the cheapest line from all the lines where the program is applicable
      */
-    _getCheapestLine() {
+    _getCheapestLine(reward) {
+        const applicableProductIds = new Set(reward.all_discount_product_ids.map((p) => p.id));
         const filtered_lines = this.getOrderlines().filter(
-            (line) => !line.comboParent && !line.reward_id && line.getQuantity
+            (line) =>
+                !line.comboParent &&
+                !line.reward_id &&
+                line.getQuantity &&
```

Comment:
```suggestion
                line.getQuantity() &&
```

Same, isn't a getter.